### ISSUE TITLE
Docs: document limitations of custom image loader

### ISSUE THREAD

**Issue by StructuralRealist** [vercel/next.js#51842]
### What is the improvement or update you wish to see?

 When creating a custom image loader, it seems that although the initial render might be server-side and has access to private environment variables, subsequent client-side renders do not. This results in the images being optimized on initial page load, but not after navigating to another page. It took me a long time to figure this out, so it would be helpful if the page about custom image loaders (https://nextjs.org/docs/pages/api-reference/next-config-js/images) mentions that you can only use public environment variables. 

### Is there any context that might help us understand?

Many image proxies like imgproxy and Thumbor use a signing mechanism for the image URL. This requires a key that you don't want to make public with NEXT_PUBLIC_. When you use a private environment variable it looks like it's working but it's only after navigating to another page that it breaks down.

### Does the docs page already exist? Please link to it.

https://nextjs.org/docs/pages/api-reference/next-config-js/images

**Comment by JesseKoldewijn** [comment]
Hey man, if I'm going to be honest this doesn't sound like a strictly speaking Next.js issue by itself. If you need to get access to images where you would need private env vars for I would recommend wrapping the image request in a vercel/og image response to allow getting the image totally on server side only. 

I'm sure there are other solutions to this issue but this is how I've been getting around issues like that.

**Comment by StructuralRealist** [comment]
Hi Jesse, 

Thanks for your reply. I don't understand why this is not a Next.js issue? Custom image loaders are a Next.js feature and the way private and public environment variables are handled is a Next.js feature too. Mind that this ticket is just for documenting what the limits are of using server-side features in the loader function, not for adding support for private variables (although that would be great of course). 

On the other hand, vercel/og is _not_ a Next.js feature as it's only available on Vercel.

**Comment by JesseKoldewijn** [comment]
> Hi Jesse, 
> 
> Thanks for your reply. I don't understand why this is not a Next.js issue? Custom image loaders are a Next.js feature and the way private and public environment variables are handled is a Next.js feature too. Mind that this ticket is just for documenting what the limits are of using server-side features in the loader function, not for adding support for private variables (although that would be great of course). 
> 
> On the other hand, vercel/og is _not_ a Next.js feature as it's only available on Vercel.

What exactly prevents you from wrapping the images on serverside while using these private env vars if I may ask? Haven't ran into this specific issue myself which is why I answered in such a wide scope.

**Comment by StructuralRealist** [comment]
I want to use the next/image component because it offers some other features than just changing the url based on the width (like handling layout shifts and srcsets). It also seems to be the recommended way to include images in Next. Custom image loaders are also the recommended way to use something like Thumbor (they even give an example of this). The problem arises because Thumbor (and imgproxy) highly recommend to use URL signing.

So you get something like `const signature = sign(SALT, path, KEY);` to add to your URL to protect agains denial-of-service attacks. The salt/key pair needs to be secret so I added private environment variables. And this worked. I refresh the page and the images are optimized. But when the `/` route is loaded, Next.js starts loading (prefetching) other routes, like `/about` and such. What I noticed was that images on _these_ pages where _not_ optimized. Even for SSR components. If I would navigate to another page and hit the back button, now suddenly all my images are unoptimized. 

I tried to figure out why, think it was a caching issue at first, but the only way to fix it was by using public environment variables (NEXT_PUBLIC_). This of course makes the protection a bit weak 😅.  Anyway, either this behavior is incorrect and this should be a bug ticket, or it is correct but then it would be nice to have a little bit more documentation about it.

**Comment by JesseKoldewijn** [comment]
Understandable,
And to be honest I agree that there could be some some possible docs improvements on cases like that. What I do recommend is using @t3-oss/env for managing your env vars for improved type-safety and also a more clear flow around what can be accessed and from where it can.

**Comment by leerob** [comment]
I'm not sure we need documentation for this 🙏
